pkg/kinflate/tree/builder.go

```diff
@@ -0,0 +1,200 @@
+	bytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	var m manifest.Manifest
+	// TODO: support json
+	err = yaml.Unmarshal(bytes, &m)
+	if err != nil {
+		return nil, err
+	}
+	dir, _ := path.Split(filename)
+	adjustpath.AdjustPathsForManifest(&m, []string{dir})
+	return &m, err
+}
+
+// validateManifestPath loads the manifest from the given path.
+// It returns ManifestData and an potential error.
+func validateManifestPath(mPath string) (string, error) {
+	f, err := os.Stat(mPath)
+	if err != nil {
+		return "", err
+	}
+	if f.IsDir() {
+		mPath = path.Join(mPath, constants.KubeManifestFileName)
+		_, err = os.Stat(mPath)
+		if err != nil {
+			return "", err
+		}
+	} else {
+		if !strings.HasSuffix(mPath, constants.KubeManifestFileName) {
+			return "", fmt.Errorf("expecting file: %q, but got: %q", constants.KubeManifestFileName, mPath)
+		}
+	}
+	return mPath, nil
+}
+
+func manifestToManifestData(m *manifest.Manifest) (*ManifestData, error) {
+	mdata := &ManifestData{}
+	var err error
+	mdata.Name = m.Name
+	mdata.NamePrefix = NamePrefixType(m.NamePrefix)
+	mdata.ObjectLabels = m.ObjectLabels
+	mdata.ObjectAnnotations = m.ObjectAnnotations
+	mdata.Resources, err = pathsToMap(m.Resources)
+	if err != nil {
+		return nil, err
+	}
+	mdata.Patches, err = pathsToMap(m.Patches)
+	if err != nil {
+		return nil, err
+	}
+	mdata.Configmaps, err = cutil.MakeMapOfConfigMap(m)
+	if err != nil {
+		return nil, err
+	}
+	mdata.Secrets, err = cutil.MakeMapOfGenericSecret(m)
+	if err != nil {
+		return nil, err
+	}
+	TLSSecrets, err := cutil.MakeMapOfTLSSecret(m)
+	err = mergemap.Merge(mdata.Secrets, TLSSecrets)
+	if err != nil {
+		return nil, err
+	}
+	return mdata, nil
+}
+
+func pathsToMap(paths []string) (map[gvkn.GroupVersionKindName]*unstructured.Unstructured, error) {
+	res := map[gvkn.GroupVersionKindName]*unstructured.Unstructured{}
+	for _, path := range paths {
+		err := pathToMap(path, res)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func pathToMap(path string, into map[gvkn.GroupVersionKindName]*unstructured.Unstructured) error {
+	_, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	if into == nil {
+		into = map[gvkn.GroupVersionKindName]*unstructured.Unstructured{}
+	}
+
+	var e error
+	filepath.Walk(path, func(filepath string, info os.FileInfo, err error) error {
+		if err != nil {
+			e = err
+			return err
+		}
+		// Skip all the dir
+		if info.IsDir() {
+			return nil
+		}
+
+		err = fileToMap(filepath, into)
+		return nil
+	})
+	return e
+}
+
+func fileToMap(filename string, into map[gvkn.GroupVersionKindName]*unstructured.Unstructured) error {
+	f, err := os.Stat(filename)
+	if err != nil {
+		return err
+	}
+	if f.IsDir() {
+		return fmt.Errorf("%q is NOT expected to be an dir", filename)
+	}
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return err
+	}
+
+	_, err = kutil.Decode(content, into)
```

Comment:
Does this complain about name collisions, i.e. name already in the map?

Reply:
No. Will add a check in `func Decode()`